http-client/src/main/java/io/airlift/http/client/GatheringByteArrayInputStream.java

```diff
@@ -0,0 +1,99 @@
+package io.airlift.http.client;
+
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkElementIndex;
+import static com.google.common.base.Preconditions.checkPositionIndex;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.lang.System.arraycopy;
+import static java.util.Objects.requireNonNull;
+
+public class GatheringByteArrayInputStream
+        extends InputStream
+{
+    private final Iterator<byte[]> buffers;
+    private final byte[] singleByte;
+    private byte[] currentBuffer;
+    private int currentBufferPosition;
+    private long remainingBytes;
+
+    public GatheringByteArrayInputStream(List<byte[]> buffers, long totalBytes)
+    {
+        checkArgument(totalBytes >= 0, "totalBytes should equal to or greater than 0");
+        this.buffers = requireNonNull(buffers, "buffers is null").iterator();
+        this.remainingBytes = totalBytes;
+
+        currentBuffer = new byte[0];
+        currentBufferPosition = 0;
+        singleByte = new byte[1];
+    }
+
+    @Override
+    public synchronized int read(byte[] buffer)
+    {
+        return read(buffer, 0, buffer.length);
+    }
+
+    @Override
+    public synchronized int read()
+    {
+        int bytes = read(singleByte);
+        if (bytes == -1) {
+            return -1;
+        }
+        return singleByte[0] & 0x000000ff;
+    }
+
+    @Override
+    public synchronized long skip(long n)
+    {
+        if (n < 0) {
+            return 0;
+        }
+        return max(doRead(null, 0, n, false), 0);
+    }
+
+    @Override
+    public synchronized int read(byte[] buffer, int offset, int length)
+    {
+        requireNonNull(buffer, "buffer is null");
+        checkElementIndex(offset, buffer.length, "offset is out of range");
+        checkPositionIndex(length, buffer.length - offset, "length is out of range");
+
+        return (int) doRead(buffer, offset, length, true);
+    }
+
+    public void close() {}
+
+    private long doRead(byte[] buffer, int offset, long length, boolean copyData)
+    {
+        checkArgument(buffer != null || !copyData, "buffer should be non-null when copying data is requested");
+
+        if (remainingBytes == 0) {
+            return -1;
+        }
+
+        long totalReadBytes = 0;
```

Comment:
I think you can get rid of this field by subtracting `readBytes` from length and adding to `offset` after every copy.  This will simplify the `while` condition  and the `min` and `arrayCopy` calls below.